realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -1581,11 +1586,16 @@ void BaseRealSenseNode::frame_callback(rs2::frame frame)
                 clip_depth(depth_frame, _clipping_distance);
             }
 
-            ROS_DEBUG("num_filters: %d", static_cast<int>(_filters.size()));
-            for (std::vector<NamedFilter>::const_iterator filter_it = _filters.begin(); filter_it != _filters.end(); filter_it++)
-            {
-                ROS_DEBUG("Applying filter: %s", filter_it->_name.c_str());
-                frameset = filter_it->_filter->process(frameset);
+//            nomagicStoreFramesetForLazyProcessing(frameset);
+            nomagic_muxer.invoke(frame);
+            bool apply_filters_now = nomagicFramesetHasSubscribers(frameset) || !nomagic_lazy_filtering;
```

Comment:
Do I understand correctly that this could produce bad temporal filtering for the first few frames after connecting to a topic? With the way we use ImageSnappers, it would mean that we would get frames without temporal filtering, unless we run some image_view in the background for debugging - then it would disappear, making it a bit of a Heisenberg bug. Thus, if my understanding is correct, it could be quite dangerous and we could:
* have separate "filters" and "nomagic_service_filter" parameters and use an empty list for the former while putting all the filters into the latter. That could still be a bit confusing, but would be self-consistent and should be easy to implement.
* wait for at least 9 frames before posting something on a topic. Then we would have no correctness bugs, but latency ones, which is easier to debug.
* Przemek, maybe some other ideas or comments?

Reply:
Yes, your understanding is correct.
> - have separate "filters" and "nomagic_service_filter" parameters and use an empty list for the former while putting all the filters into the latter. That could still be a bit confusing, but would be self-consistent and should be easy to implement.
Can you explain more thoroughly? I think I'm missing some understanding here.
> wait for at least 9 frames before posting something on a topic. Then we would have no correctness bugs, but latency ones, which is easier to debug.
What would the 9 frames change here? It will indeed fill up the temporal filter's state, but after a while, we'd have outdated frames and thus artifacts after filtering.

Reply:
Note that from the point of view of the ros-realsense, we don't know when ImageSnapper will decide to subscribe to get a single frame. Therefore to publish afterwards a correctly temporal-filtered frame, we need to constantly feed the temporal filter (which happens in the original ros-realsense version) as the frames arrive.

I believe that the actual problem here lies in the ImageSnapper which uses topics in a wrong (RPC-ish) way. I see two possible ways to correct this:
- Make ImageSnapper subscribe continuously to the topic to force and cache the latest image - GIL-risky solution
- Prevent topic-based ImageSnappers from using ros-realsense with nomagic_lazy_filtering==true. This can be enforced by just removing the topics when nomagic_lazy_filtering==true. I can add nomagic_lazy_filtering to be configurable per camera (container) (in /etc/nomagic/realsense-config.yml).

Reply:
@mikolajz We can also leave topics, use a separate filters object to avoid modifying the temporal filter state by both service and topic paths and make nomagic_lazy_filtering be disabled by default and document well what are the consequences of enabling it (that topics should no longer be used for a single frame capture).